[ASTGen] Syntax-tree based interface hash

rintaro opened the pull request:
Added `StableHasher` in ASTGen. Mostly copied from `stdlib/public/core/Hasher.swift` and ``stdlib/public/core/SipHash.swift``

For now, interface hash / fingerprint are emitted during the AST generation, but it should really be a separate request. I will separate fingerprint things from `ParseSourceFileRequest` and `ParseMembersRequest` in followups

https://github.com/swiftlang/swift/issues/76527
rdar://117150870

rintaro committed:
[ASTGen] Add syntax tree based Fingerprint

rintaro committed:
[ASTGen] Use syntax tree based interface hash

rintaro commented:
@swift-ci Please smoke test

CodaFi commented on lib/ASTGen/Sources/ASTGen/StableHasher.swift:
```diff
@@ -0,0 +1,312 @@
+    private var v2: UInt64 = 0x6c7967656e657261
+    private var v3: UInt64 = 0x7465646279746573
+    // The fields below are reserved for future use. They aren't currently used.
+    private var v4: UInt64 = 0
```
I think we can delete these.

rintaro commented:
@swift-ci Please smoke test

rintaro committed:
[ASTGen] Use arch(_:) instead of _pointerBitWidth(_:)

We still need to support older Swift compilers which does't have
'_pointerBitWidth(_:)' pletform condition.

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test Linux
